### ISSUE TITLE
A way to cut off shape generation at certain points

### ISSUE THREAD

**Issue by AArnott** [eiriktsarpalis/PolyType#85]
I was reviewing the generated code from `[GenerateShape]`, and found several classes to support types that don't get automatically serialized. I believe it's because certain properties are typed with types that I don't expect to be automatically serializable (at least not in an optimal representation) so I have custom converters for them. But the type shape remains in the generated form, needlessly bloating the software. (Will unused type shapes be trimmed?)

I wonder if we should have an attribute that I can apply to types or properties that reference those types to hint to the source generation system that it should not follow this type reference path any further.

Thoughts?

**Comment by eiriktsarpalis** [comment]
Does using [PropertyShapeAttribute.Ignore](https://eiriktsarpalis.github.io/PolyType/api/PolyType.PropertyShapeAttribute.html#PolyType_PropertyShapeAttribute_Ignore) prevent generation of the nested type?

**Comment by AArnott** [comment]
Yes, that works. Though that also removes the property from its declaring type's shape, so it won't be serialized even by my custom converter.

**Comment by eiriktsarpalis** [comment]
On the type with the custom converter itself, you could try applying the new TypeShapeAttribute with TypeShapeKind set to None.

**Comment by AArnott** [comment]
In my cases, I don't declare the type that I don't want the shape for. (e.g. Stream, cryptography keys, etc.)

This isn't a blocking issue, so feel free to Won't Fix. It's more an observation about some downside, but the 'fix' may be worse...

**Comment by AArnott** [comment]
> > On the type with the custom converter itself, you could try applying the new TypeShapeAttribute with TypeShapeKind set to None.
> In my cases, I don't declare the type that I don't want the shape for. (e.g. Stream, cryptography keys, etc.)

This should become possible with #128 plus the work to add `TypeShapeKind` to `TypeShapeExtension`:

```cs
[assembly: TypeShapeExtension(typeof(ExternalType), Kind = TypeShapeKind.None)]
```

**Comment by AArnott** [comment]
Although I wonder how the Reflection provider would discover this. For AssociatedTypes it was fine because with reflection we don't need them. But if the idea is `Kind` can be set to arbitrary values through the extension attribute, how would Reflection discover it?

Also, there is a possibility of the same type being extended multiple times. For AssociatedTypes, these can simply be unioned together, so no conflicts. But for `Kind`, if it is set on multiple of these attributes to conflicting values, we'd need a resolution policy. 
For source generation, the local assembly can 'win' over referenced assemblies. But what about reflection?

**Comment by eiriktsarpalis** [comment]
> Although I wonder how the Reflection provider would discover this. For AssociatedTypes it was fine because with reflection we don't need them. But if the idea is `Kind` can be set to arbitrary values through the extension attribute, how would Reflection discover it?

Presumably the same way like it would discover associated types? Presumably walk all assemblies in the current domain for attribute annotations and weave this information in _before_ any types get generated.

**Comment by AArnott** [comment]
I have more thoughts on solving the reflection provider problem over at https://github.com/eiriktsarpalis/PolyType/issues/118#issuecomment-2721375156. Perhaps let's continue the discussion there.

**Comment by AArnott** [comment]
While type shape extension support for setting Kind is interesting, setting it to None seems a bit narrow-sighted now that I think about it. Even if my serializer doesn't need a shape for a particular type because I have a custom converter, if I set it to None via type shape extension in my serializer, that breaks the shape for any other polytype-based library that might want its original shape.

So I'll close this issue.
